src/v/ssx/tests/when_all_test.cc

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "container/fragmented_vector.h"
```

Comment:
I think it's worth adding some basic tests for `future<>`, as well as move-only types (e.g., `std::unique_ptr<int>`).

Reply:
Will add some testing for move-only types :+1: 

This version cannot handle `future<>`. The equivalent overload for `void` futures is `future<> when_all_succeed(std::vector<future<>>&&)` and requires separate handling as it doesn't return a range, but a void future. Do you think it is worth adding?

Reply:
> This version cannot handle `future<>`. The equivalent overload for `void` futures is `future<> when_all_succeed(std::vector<future<>>&&)` and requires separate handling as it doesn't return a range, but a void future. Do you think it is worth adding?

It's a common usage pattern, but it can always be added later.